### PR TITLE
hive: Rebase on cilium/hive

### PR DESCRIPTION
**Pull request by bimmlerd**
The general idea is to have `pkg/hive` contain the cilium-specific adaptations of `cilium/hive`. Specifically, that's providing certain cells like statedb and jobs directly, and having module decorators creating the right health and logging components.

It's one big commit as it has proven difficult to get intermediate commits to compile (which our CI mandates). The health part needs a bit of love after this, we can probably move `healthv2` into `hive/health` as it's strongly tied to cilium's hive.

The bulk of the changes are mechanical. The categories are:

1. import path change from `cilium/cilium/pkg/hive/{cell,job}` to `cilium/hive/{cell,job}`
2. giving `hive.{Start,Stop,Run}` an `slog.Logger` to provide to the hive,
3. changes to the health interfaces - `Scope` and `HealthReporter` are unified under just `Health`
4. modules are provided with a job group scoped to their module id directly, instead of having to create a group from a registry

### CODE REVIEW

File: pkg/hive/hive.go [hunk 1]
```diff
@@ -4,399 +4,101 @@
...
+		},
+	}
+	return upstream.NewWithOptions(
+		upstream.Options{
+			Logger:                 logging.DefaultSlogLogger,
+			EnvPrefix:              "CILIUM_",
+			ModulePrivateProviders: []cell.ModulePrivateProvider{jobGroupProvider},
+			ModuleDecorators:       moduleDecorators,
+			DecodeHooks:            decodeHooks,
+			StartTimeout:           10 * time.Minute,
```
Comment (joamaki) [inline pkg/hive/hive.go]: Let's keep the 5 minute start timeout and 1 minute stop timeout.

File: pkg/hive/hive.go [hunk 2]
```diff
@@ -4,399 +4,101 @@
...
+			StopTimeout:            10 * time.Minute,
+		},
```
Comment (joamaki) [inline pkg/hive/hive.go:73]: Let's set the `LogThreshold` option that Joe added. 100ms.

**Review by ti-mo — APPROVED** [review]

File: operator/pkg/controller-runtime/cell.go [hunk 3]
```diff
@@ -51,10 +51,10 @@ func newScheme() (*runtime.Scheme, error) {
...
 	cell.In
 
-	Logger      logrus.FieldLogger
-	Lifecycle   cell.Lifecycle
-	JobRegistry job.Registry
-	Scope       cell.Scope
+	Logger    logrus.FieldLogger
+	Lifecycle cell.Lifecycle
+	JobGroup  job.Group
+	Scope     cell.Health
```
Comment (mhofstetter) [inline operator/pkg/controller-runtime/cell.go]: nit: there are some leftovers of `Scope cell.Health` that probably should be renamed to `Health`.

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -369,7 +370,7 @@ func (c *controller) getLogger() *logrus.Entry {
 
 // recordError updates all statistic collection variables on error
 // c.mutex must be held.
-func (c *controller) recordError(err error, hr cell.HealthReporter) {
+func (c *controller) recordError(err error, hr cell.Health) {
```
Comment (mhofstetter) [inline pkg/controller/controller.go]: nit: `hr` -> `health` / `h` (same below in `recordSuccess`)

File: pkg/envoy/cell.go [hunk 5]
```diff
@@ -214,9 +216,10 @@ type versionCheckParams struct {
 	cell.In
 
 	Lifecycle        cell.Lifecycle
+	Slog             *slog.Logger
```
Comment (mhofstetter) [inline pkg/envoy/cell.go:218]: not a fan of having two loggers injected here and in other places (slog & logrus). Is it up to the owners to clean this up afterwards?

File: pkg/node/manager/manager.go [hunk 6]
```diff
@@ -136,7 +137,7 @@ type manager struct {
 	controllerManager *controller.Manager
 
 	// healthScope reports on the current health status of the node manager module.
-	healthScope cell.Scope
+	healthScope cell.Health
```
Comment (mhofstetter) [inline pkg/node/manager/manager.go]: nit: `healthScope` -> `health`

File: daemon/restapi/api_limits_test.go [hunk 7]
```diff
@@ -13,8 +13,11 @@ import (
 	"github.com/stretchr/testify/assert"
 	timeRate "golang.org/x/time/rate"
 
+	"github.com/cilium/hive/hivetest"
+
```
Comment (gandro) [inline daemon/restapi/api_limits_test.go]: nit: unnecessary empty line

File: pkg/k8s/client/client_test.go [hunk 8]
```diff
@@ -11,12 +11,14 @@ import (
 	"time"
 
 	. "github.com/cilium/checkmate"
+	"github.com/cilium/hive/hivetest"
 	"github.com/spf13/pflag"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	"github.com/cilium/hive/cell"
```
Comment (gandro) [inline pkg/k8s/client/client_test.go:19]: General nit/question (applies to all files, but here is a good example where the discrepancy shows): For other packages under the cilium GitHub organization (e.g. cilium/dns) we don't introduce a new group in the import list. Here it's even inconsistent between `cilium/hive/cell` and `cilium/hive/hivetest`.

IMO, the `github.com/cilium/hive/cell` import should belong to the third-party group as all other dependencies in `vendor/`.

**Review by gandro — COMMENTED** [review]
The changes in my CODEOWNERS look good mostly, thanks! My biggest question/feedback is with regards to the package import grouping

File: pkg/k8s/client/client_test.go (hunk 8 above)
Comment (gandro) [reply pkg/k8s/client/client_test.go:19]: For context, here's the ordering we decided on a few years ago - https://github.com/cilium/cilium/pull/18357 (c.f. https://github.com/cilium/cilium/pull/18357/files#r777600306)

I'm surprised this isn't enforced by any CI workflow to be honest

**Review by nebril — APPROVED** [review]
ci-structure LGTM. Good luck getting it merged!

File: daemon/restapi/api_limits_test.go (hunk 7 above)
Comment (bimmlerd) [reply daemon/restapi/api_limits_test.go]: The canonical `goimports -w -local github.com/cilium/cilium` does not change this. If I manually clean up the space it's also happy, but I'm not super excited to discuss formatting if it's not tool-enforced.

Don't get me wrong, happy to clean this one up, but in general my standpoint is mostly "if the linters don't complain, it's good enough"

File: pkg/k8s/client/client_test.go (hunk 8 above)
Comment (bimmlerd) [reply pkg/k8s/client/client_test.go:19]: https://github.com/cilium/cilium/pull/18357#discussion_r777612575 I guess says it best, I really don't want to spend time manually fixing up stuff

File: pkg/envoy/cell.go (hunk 5 above)
Comment (bimmlerd) [reply pkg/envoy/cell.go:218]: Yeah, bit of context here:

1. the transition logrus -> slog starts with this PR, but it's hidden for most downstream modules, because `hive/job` is the only consumer at this time. Since we also inject `job.Group` for modules now, most modules don't need to create new groups from the registy, and hence never see that the logger is now slog.
2. In this case, however, it looked like multiple groups were spawned with some thought put into it, and I didn't want to mess that up - hence I "just" passed in the new slog Logger needed to create the groups. This will be cleaned up once the whole logrus -> slog transition is done, because then the logrus logger will disappear completely
3. Who will do the logrus -> slog I don't think is clear yet, but doing it as part of this PR is out of scope.

Comment (mhofstetter) [reply pkg/envoy/cell.go:218]: Thanks for the explanation David!

**Review by mhofstetter — APPROVED** [review]

**Review by joamaki — APPROVED** [review]

File: operator/pkg/controller-runtime/cell.go (hunk 3 above)
Comment (bimmlerd) [reply operator/pkg/controller-runtime/cell.go]: ack, fixing all occurrences.

File: pkg/k8s/client/client_test.go (hunk 8 above)
Comment (gandro) [reply pkg/k8s/client/client_test.go:19]: Yeah, so it seems that our tooling is insufficient :disappointed: 

I'm pretty sure there are probably already a bunch of already broken import statements in the tree because of that.

However, I still feel uneasy about merging this PR as is, since it objectively does introduce more inconsistencies, even if our linters don't complain about it. I did a quick test with `gofumpt` and it also does not seem to pick up the broken imports :/

Comment (bimmlerd) [reply pkg/k8s/client/client_test.go:19]: Yeah, fair enough. I'll hack something together with duct tape and sed to fix this up :safety_vest:

Comment (gandro) [reply pkg/k8s/client/client_test.go:19]: Did some more googling and at least on this file here the following hack seems to do the trick: https://gist.github.com/soniah/c11633551c6dd84dab66cad20453cfa8

We probably want to add that to our CI if it work reliably

Comment (bimmlerd) [reply pkg/k8s/client/client_test.go:19]: yep. that particular iteration is very slow, as it spawns goimports once per file, but could speed that up by deleting all newlines in import first, and then `goimports -w -local github.com/cilium/cilium ./' all the files. Note that some of the k8s pkg and generated code is not actually formatted correctly, so it's a bit of work :sweat:

EDIT: also, should've probably run it only on changed files, there's a bunch of new changes now, oh well

File: pkg/fqdn/dns/dns.go [hunk 9]
```diff
@@ -12,7 +12,6 @@ package dns
 import "strings"
 
 // These functions were copied and adapted from github.com/miekg/dns.
-
```
Comment (gandro) [inline pkg/fqdn/dns/dns.go]: The script was a bit over-eager when removing empty lines from imports here, presumably because there is only one import statement :grimacing:

File: pkg/fqdn/proxy/ipfamily/ipfamily.go [hunk 10]
```diff
@@ -6,11 +6,10 @@ package ipfamily
 import "golang.org/x/sys/unix"
 
 type IPFamily struct {
-	Name       string
-	UDPAddress string
-	TCPAddress string
-	Localhost  string
-
```
Comment (gandro) [inline pkg/fqdn/proxy/ipfamily/ipfamily.go]: The script was a bit over-eager when removing empty lines from imports here

File: pkg/ipam/node_test.go [hunk 11]
```diff
@@ -13,7 +13,6 @@ type testNeededDef struct {
 	maxallocate int
 	result      int
 }
-
```
Comment (gandro) [inline pkg/ipam/node_test.go]: The script was a bit over-eager when removing empty lines from imports here

File: pkg/ipam/node_test.go [hunk 12]
```diff
@@ -35,7 +34,6 @@ var neededDef = []testNeededDef{
 	{8, 4, 8, 0, 8, 0},
 	{4, 4, 8, 0, 8, 4},
 }
-
```
Comment (gandro) [inline pkg/ipam/node_test.go]: The script was a bit over-eager when removing empty lines from imports here

File: pkg/nodediscovery/cell.go [hunk 13]
```diff
@@ -3,14 +3,13 @@
...
 
-import "github.com/cilium/cilium/pkg/hive/cell"
+import "github.com/cilium/hive/cell"
 
 // The node discovery cell provides the local node configuration and node discovery
 // which communicate changes in local node information to the API server or KVStore.
 var Cell = cell.Module(
 	"nodediscovery",
 	"Communicate changes in local node information to the API server or KVStore",
-
```
Comment (gandro) [inline pkg/nodediscovery/cell.go]: The script was a bit over-eager when removing empty lines from imports here

**Review by gandro — COMMENTED** [review]
Thanks so much for cleaning up the imports! I did find some instances in my CODEOWNERs where the imports cleanup went slightly wrong (I have not checked files not owned by me)

File: pkg/fqdn/dns/dns.go (hunk 9 above)
Comment (bimmlerd) [reply pkg/fqdn/dns/dns.go]: ahh crap. yeah let me look through those :disappointed:

**Review by gandro — APPROVED** [review]
Awesome, this looks good to me now :rocket:

**Review by chancez — APPROVED** [review]
Hubble related changes lgtm.

File: daemon/restapi/api_limits_test.go [hunk 14]
```diff
@@ -9,39 +9,41 @@ import (
...
 
 	"github.com/cilium/cilium/pkg/hive"
-	"github.com/cilium/cilium/pkg/hive/cell"
 	"github.com/cilium/cilium/pkg/rate"
 )
 
 // TestRateLimiterConfigFlag checks that rateLimiterConfig is properly parsed from
 // command-line flag.
-func testRateLimiterConfig(t *testing.T, setConfig func(h *hive.Hive, limiter string, limits string)) {
+func testRateLimiterConfig(t *testing.T, burst int, setConfig func(h *hive.Hive, limiter string, limits string)) {
```
Comment (learnitall) [inline daemon/restapi/api_limits_test.go:24]: The changes to the burst parameter aren't documented in the commit message or the PR. Are these changes related? Or accidentally lumped in?

**Review by learnitall — COMMENTED** [review]
Woohoo! :partying_face:. Quick clarifying question for you, but otherwise LGTM for API changes.

**Comment by learnitall** [comment]
I did a quick sanity check and I think there are still some places where the old hive package is referenced. Were these left in on purpose? It looks like there are some leftover references in Documentation and comments, as well as some leftover imports to `github.com/cilium/cilium/pkg/hive`:

```bash
$ grep 'pkg/hive' -R . --include \*.go --include \*.rst | wc -l
89
```

Examples:

```
./Documentation/contributing/development/hive.rst:`hive.New <https://pkg.go.dev/github.com/cilium/cilium/pkg/hive#New>`_ constructor. 
./vendor/github.com/cilium/hive/hive.go:// See pkg/hive/example for a runnable example application.
./pkg/proxy/logger/logger_test.go:	"github.com/cilium/cilium/pkg/hive"
```

File: daemon/restapi/api_limits_test.go (hunk 14 above)
Comment (bimmlerd) [reply daemon/restapi/api_limits_test.go:24]: It's a broken test, due to shared global state in the pflag library. Will add a note to the commit message.

**Comment by bimmlerd** [comment]
> I did a quick sanity check and I think there are still some places where the old hive package is referenced. Were these left in on purpose? It looks like there are some leftover references in Documentation and comments, as well as some leftover imports to `github.com/cilium/cilium/pkg/hive`:
> 
> ```shell
> $ grep 'pkg/hive' -R . --include \*.go --include \*.rst | wc -l
> 89
> ```
> 
> Examples:
> 
> ```
> ./Documentation/contributing/development/hive.rst:`hive.New <https://pkg.go.dev/github.com/cilium/cilium/pkg/hive#New>`_ constructor. 
> ./vendor/github.com/cilium/hive/hive.go:// See pkg/hive/example for a runnable example application.
> ./pkg/proxy/logger/logger_test.go:	"github.com/cilium/cilium/pkg/hive"
> ```

These are intentional. The PR's purpose is not to remove `pkg/hive`, but to use the implementation of `hive` in `github.com/cilium/hive`. `cilium/hive` is meant to be usable for projects unrelated to cilium, whereas `github.com/cilium/cilium/pkg/hive` is the adaptation of `hive` for the purposes of the cilium agent (see PR description/commit message). `pkg/hive/job` and `pkg/hive/cell` are removed from `cilium/cilium`, hence these import paths had to be rewritten, but not `pkg/hive`, as, for example, `pkg/hive.New` is how we inject our default cells into the hive.

**Review by dylandreimerink — APPROVED** [review]

**Review by harsimran-pabla — APPROVED** [review]

**Review by learnitall — APPROVED** [review]
Looks good, thank you! My apologies for the oversite there :facepalm:

**Review by aanm — APPROVED** [review]
